Adds a tab to pull repositories into the hub

AlexIsBadWithComputers opened the pull request:
**What this Extension Does**

This extension makes it easy for teachers and other users with very little technical skills but who want to see / use other notebooks they’ve found on github in their own Callysto hub environment. This is inspired by the graphic that was displayed at JupyterCon that there were over 2.5million notebooks and growing in Github (https://github.com/parente/nbestimate). 

This notebook extension creates a new tab in your Jupyterhub home/dashboard which is essentially a wrapper for nbgitpuller. In the tab, anyone can enter a github url for an external notebook that they are interested in seeing and then the extension creates an nbgitpuller link, and opens it in a new tab. From there it becomes standard 'nbgitpulling' where the repository is synced, and opened in their hub upon completion. 

**Notes**
There is input sanitation that addresses the following 

1. Trying to pull non-github links 
2.  Trying to pull from links that 404. 
3. Trying to pull github links that aren't repositories
       a. I'm sure there are some edge cases where this fails
4.  I've done my best to make sure the links provided to nbgitpuller don't fail, but I'm sure there are cases where it still will.  
5. This may cause some users to git their 1GB disk limit very quickly.

ianabc committed:
Debian testing stack

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into debian-stretch

ianabc committed:
Merge branch 'master' into debian-stretch

ianabc committed:
Dont conda update, it will bring in the broken R package

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into debian-stretch

ianabc committed:
Adding upstream changes

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into debian-stretch

Signed-off-by: Ian Allison <iana@pims.math.ca>

Conflicts:
	datascience-notebook/Dockerfile

ianabc committed:
Debian testing version. Playing with julia 0.5

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into debian-stretch

ianabc committed:
Importing some stuff from juliabox

Signed-off-by: Ian Allison <iana@pims.math.ca>

Commit:
Try to avoid being NB_USER specific

Signed-off-by: root <root@cybera.syzygy.ca>

ianabc committed:
Add crypto dependencies and a pip.conf

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Forgot to add the pip.conf

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
notebook_config added and julia package list updated

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge remote-tracking branch 'upstream/master' into merge-us

Signed-off-by: Ian Allison <iana@pims.math.ca>

Conflicts:
	Makefile
	base-notebook/Dockerfile

ianabc committed:
Merge branch 'master' into ds2

Conflicts:
	base-notebook/Dockerfile

ianabc committed:
Merge branch 'ds2' into ds2-test

ianabc committed:
Update startup script to match jupyterhub version

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into ds2-test

Conflicts:
	datascience-notebook/Dockerfile

ianabc committed:
Stretch please

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Default now doesnt use the output filtering hook

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into ds2a

Signed-off-by: Ian Allison <iana@pims.math.ca>

Conflicts:
	base-notebook/Dockerfile
	datascience-notebook/Dockerfile
	scipy-notebook/Dockerfile

ianabc committed:
Remove python2 and add google drive extension

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into ds2a

ianabc committed:
Fixed systemuser.sh for new versions of dockerspawner

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
R fixed a lot of package updates applied

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding some julia stuff

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
jupyterlab is at 0.29 now

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Use 0.8.1 weird incompatibility for 0.8.0

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Problem with julia PyPlot not using correct libz fixed by hack

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' of https://github.com/jupyter/docker-stacks into ds2a

ianabc committed:
Merge branch 'master' into ds2b

ianabc committed:
Working on factoring out pims-notebook into pims-minimal
Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Factoring out generic stuff into minimal

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding a README.md for minimal

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding a pims-r notebook

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding information about compiler requirments bug

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Mostly changes in layout of dockerfile commands

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding an notebook with nbgitpuller

This will probably be moved into the production container eventually

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Make nbgitpuller available everywhere its lovely

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'master' into ds2b

ianabc committed:
julia is used in addition to R

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding lxml and astropy

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding cars for Marie

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Mostly R related stuff for user package installation

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Add new packages, update base image

Merge branch 'master' into ds2b

Signed-off-by: Ian Allison <iana@pims.math.ca>

Conflicts:
	base-notebook/Dockerfile
	scipy-notebook/Dockerfile

ianabc committed:
Fix a type and update names for build dependencies

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Lets use the jupyter name for the moment

Signed-off-by: Ian Allison <iana@pims.math.ca>

jtopjian committed:
Fixes to the ianabc branch for building container images

jtopjian committed:
$USER and $USER_ID fixes

jtopjian committed:
Modifications to get pims images to pass tests

jtopjian committed:
Merge pull request #1 from callysto/ianabc-fixes

Fixes to the ianabc branch for building container images

jtopjian committed:
Push Travis builds to DockerHub

jtopjian committed:
Merge pull request #2 from callysto/dockerhub-test

Docker Hub Integration

jtopjian committed:
Fixing deploy typos

jtopjian committed:
Merge pull request #6 from callysto/dockerhub-test

Fixing deploy typos

jtopjian committed:
Adding swift image

ianabc committed:
Add latimes-calculate for Byron

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Remainder of packages requested by Byron

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge pull request #8 from callysto/byron-latimes-calculate

Travis checks passed so I'll merge this and start working with Byron to confirm the new packages do what he needs.

precurse committed:
Using `$HOME` variable instead of /home/

precurse committed:
Merge pull request #9 from callysto/nb_user_homedir

Use `HOME` environment variable with start.sh script inside base-notebook

jtopjian committed:
Merge pull request #7 from callysto/swiftfs

Adding swift image

Commit:
Point to new s3contents repo

jtopjian committed:
Merge pull request #12 from callysto/swiftfs-update

Point to new s3contents repo

ianabc committed:
Merge branch 'ianabc' of https://github.com/callysto/docker-stacks into ianabc

* 'ianabc' of https://github.com/callysto/docker-stacks:
  Point to new s3contents repo
  Using `$HOME` variable instead of /home/
  Remainder of packages requested by Byron
  Add latimes-calculate for Byron
  Adding swift image
  Fixing deploy typos
  Push Travis builds to DockerHub
  Modifications to get pims images to pass tests
  $USER and $USER_ID fixes
  Fixes to the ianabc branch for building container images

Commit:
Tagging images with git commit

jtopjian committed:
Merge pull request #13 from callysto/docker-tag

Tagging images with git commit

ianabc committed:
FIX wrong docker user referenced

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Recreating ordinary-user PR

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Remove unused CMD and don't bother setting matplotlibrc

matplotlibrc will be overridden by the user homedir mount anyway

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Only apply non superuser tests to pims-minimal and pims-r

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge pull request #16 from callysto/ordinary-user-3

Ordinary user 3

Commit:
Removing Travis

jtopjian committed:
Merge pull request #17 from callysto/remove-travis

Removing Travis

Commit:
Test/Verification support

Commit:
Adding scrapy and fastkml to pims-minimal

jtopjian committed:
Merge pull request #18 from callysto/scrapy

Scrapy Support

Commit:
adds tab to pull git repositories

Commit:
added more input sanitation

Commit:
fixed instructions so they'll actually work

Commit:
updates to deal with new browser security

nickurak commented on gitpuller_wrapper/main.js:
```diff
@@ -0,0 +1,141 @@
+
+            // TODO: Probably grab the hub.callysto bit from the Jupyter info 
+            // so this is general to any hub?
+            var url_head = "https://hub.callysto.ca/jupyter/hub/user-redirect/git-pull?";
```
This seems to work for my local jupyter instance -- I don't have a hub I can play with extensions on yet, but it'd sure be convenient if just a relative URL would take care of this:

```
-            var url_head = "https://hub.callysto.ca/jupyter/hub/user-redirect/git-pull?";
+            var url_head = "git-pull?";
```

AlexIsBadWithComputers replied:
Whoops! Sorry just saw this.

I agree, a relative link would be better, and we can probably get that right from the environment. Currently the main issue is with the recent security updates this can no longer give github a poke to ensure there's no typos in user URLs. Once I get that worked out I'll start digging and update this to relative links. That said, in its current state, this is kind of broken in the sense that it works - but it can no longer check if the repo exists or not.

ianabc committed:
Merge remote-tracking branch 'upstream/master' into ianabc-jupyter-upstream

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Realign with juypyter/docker-stacks up to scipy-notebook

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
set BASE_CONTAINER as an argument install vim and remove debian stuff

BASE_CONTAINER is set as an argument to docker build similar to the rest
of the jupyter/docker-stacks images. This allows us to do e.g. `make
build/pims-r DARGS="--build-arg BASE_CONTAINER=callysto/scipy-notebook"`
to select a different base image.

The jupyter/base-image sets nano as the editor, but I cry if I can't
find vim, so here it is.

ffmpeg is included in this release of ubuntu so we don't need the
deb-multimedia repository.

Tests pass with minor warnings.

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Bump R to 3.5

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Adding fizzywuzzy tqdm and stats_can conda packages

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
No need to pin pandas any more

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Remove unused files

Signed-off-by: Ian Allison <iana@pims.math.ca>

Commit:
fixes, pre commit to test on hub

Commit:
modify for user

Commit:
cleanup

Commit:
final fixes for repo

Commit:
final clean up

Commit:
updated installation instructions

Commit:
final flourish

Commit:
name changes and fixes based on code review

AlexIsBadWithComputers commented on gitpuller_wrapper/README.md:
```diff
@@ -13,10 +13,14 @@ This notebook extension (though maybe not _technically_ a notebook extension) ad
 1. From any directory, clone this repository
-2. Run `jupyter nbextension install gitpuller_wrapper --user`
-3. Run `jupyter nbextension enable gitpuller_wrapper/main  --section='common'`
+2. Enter `docker-stacks/gitpuller_wrapper` on the `gitpulltab` branch. 
```
I have removed the reference to the branch.

AlexIsBadWithComputers commented on gitpuller_wrapper/gitpuller_wrapper/__init__.py:
```diff
@@ -0,0 +1,51 @@
+from notebook.utils import url_path_join
+import requests
+
+class PingyBoi(IPythonHandler):
```
Fair enough, I have changed it to PingHandler instead.

AlexIsBadWithComputers commented on gitpuller_wrapper/gitpuller_wrapper/static/main.js:
```diff
@@ -0,0 +1,129 @@
+            // run the nbgitpuller stuff
+
+            // UPDATE: Now we use a server side python do-hickey instead to make it work
+            // now we don't have to worry about browser security updates. 
```
I'll take a look at changing some commit messages, but I've removed that comment from the file as well.

ianabc committed:
Merge pull request #22 from callysto/ianabc-jupyter-upstream

I've tried this image on some external hosts and it seems to be working fine.

Commit:
New style wildcard addresses for jupyter_notebook_config

Signed-off-by: Callysto Sysadmin <sysadmin@callysto.ca>

Commit:
Set XDG_CACHE_HOME to jupyter users cache dir for consitency

Without this `pip --user` will try to install use a location owned by
jovyan and will exit with permission denied messages.

Signed-off-by: Callysto Sysadmin <sysadmin@callysto.ca>

Commit:
Correct timezone handling for R kernel

See https://github.com/jupyter/docker-stacks/issues/502 for details. The
solution is the same as suggested there (install tzdata).

Signed-off-by: Callysto Sysadmin <sysadmin@callysto.ca>

ianabc committed:
Merge pull request #23 from callysto/xdg-fix

Set XDG_CACHE_HOME to jupyter users cache dir for consistency and add a timezone DB for R. 
Without the timezone DB R refuses to read CSVs (because it can't handle localized dates + times)

ianabc commented:
Sorry for being so slow with this. @AlexIsBadWithComputers Could you wrap this extension up as an external repository (e.g. callysto/gitpuller-wrapper). We can then install the extension by adding only the install commands to docker stacks (this is how we handle the other extensions). This will also allow external groups to install your extension independent from docker-stacks.

AlexIsBadWithComputers commented:
@ianabc 

Not a bad idea,  I've moved these files to the following repository: https://github.com/callysto/gitpuller-wrapper

Commit:
Explicitly list nbconvert to get updated version

This is to address a 500 error when downloading notebooks via nbconvert.
See [CAN-91](https://jira.cybera.ca/browse/CAN-91) or
[notebook/issues/3769](https://github.com/jupyter/notebook/issues/3769)
for more details.

Signed-off-by: Callysto Sysadmin <sysadmin@callysto.ca>

ianabc committed:
Merge pull request #24 from callysto/nbconvert-fix

Explicitly list nbconvert to get updated version.

ianabc committed:
gitpuller wrapper moved to a new repo

See [gitpuller-wrapper](https://github.com/callysto/gitpuller-wrapper)
for the nbextension. This branch will (soon) install from there.

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc committed:
Merge branch 'ianabc' into gitpulltab

ianabc committed:
Install from github master branch

Signed-off-by: Ian Allison <iana@pims.math.ca>

ianabc commented:
@AlexIsBadWithComputers  I trying to insta the master branch from github (via pip) but I think I'm missing something.  It is reporting as being installed OK
```
!jupyter nbextension list
Known nbextensions:
  config dir: /opt/conda/etc/jupyter/nbconfig
    notebook section
      gitpuller_wrapper/main  enabled 
      - Validating: OK
...
```
but I don't get the new tab, and I can't see the javascript being loaded on the menu/tree page. I'll play around with it later, but do you have any idea what I might be missing?

AlexIsBadWithComputers commented:
I think the problem is that yours is getting installed in the wrong section. Your output for gitpullerwrapper should look like this:

```
config dir: /home/jupyter/.jupyter/nbconfig
    common section
      gitpuller_wrapper/main  enabled
      - Validating: OK
```

Where as yours seemed to have gone to the default notebook section. I'm not sure why it's doing this for you, but I never tried with a pip install so that could be part of the issue. Does this work if you run the install instructions manually?

Regardless I'll do some digging (probably not until Thursday) but a quick fix might be to add a `--section='common'` to the install command in setup.py here https://github.com/callysto/gitpuller-wrapper/blob/master/setup.py#L39 .

Although, I thought I didn't need that any more and it was working without it, but perhaps pip is doing something different than a manual install?

AlexIsBadWithComputers commented:
@ianabc 

Whoops! Sorry, the `--section='common'` flag would need to be on the enable command here: 

https://github.com/callysto/gitpuller-wrapper/blob/master/setup.py#L40

NOT the install command

ianabc commented:
Tidying up, if this is still needed please let me know and I can take another look
